File: clinical-hms-api/app/schemas/triage.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.db.models.visit import TriagePriority


class VitalsCreate(BaseModel):
    temperature: float | None = Field(default=None, ge=30.0, le=45.0)
    bp_systolic: int | None = Field(default=None, ge=50, le=300)
    bp_diastolic: int | None = Field(default=None, ge=30, le=200)
    heart_rate: int | None = Field(default=None, ge=20, le=300)
    oxygen_saturation: int | None = Field(default=None, ge=0, le=100)
    respiratory_rate: int | None = Field(default=None, ge=0, le=100)
    weight_kg: float | None = Field(default=None, ge=0.5, le=500.0)


class VitalsResponse(BaseModel):
    id: int
    visit_id: int
    temperature: float | None
    bp_systolic: int | None
    bp_diastolic: int | None
    heart_rate: int | None
    oxygen_saturation: int | None
    respiratory_rate: int | None
    weight_kg: float | None
    recorded_by_id: int
    recorded_at: datetime

    model_config = {"from_attributes": True}


class TriagePrioritySet(BaseModel):
    priority: TriagePriority


class TriageQueueItem(BaseModel):
    visit_id: int
    patient_id: int
    folder_number: str
    full_name: str
    reason_for_visit: str
    triage_priority: TriagePriority | None
    checked_in_at: datetime
    wait_minutes: int
    has_vitals: bool

    model_config = {"from_attributes": True}

File: clinical-hms-web/src/types/triage.ts
/**
 * Triage domain types — mirrors the backend triage schemas.
 *
 * TriagePriority represents the South African Triage Scale (SATS) colour
 * categories. The order in TRIAGE_PRIORITIES (lib/triageValidation.ts) matches
 * clinical severity from least to most critical.
 *
 * VitalsCreate field names must exactly match the API VitalsCreate Pydantic
 * schema (app/schemas/triage.py). A mismatch causes a 422 Validation Error
 * from FastAPI.
 */

export type TriagePriority = 'green' | 'yellow' | 'orange' | 'red'

export type TriageQueueItem = {
  visit_id: number
  patient_id: number
  folder_number: string
  full_name: string
  reason_for_visit: string
  triage_priority: TriagePriority | null
  checked_in_at: string
  wait_minutes: number
  has_vitals: boolean
}

// Matches the API VitalsCreate schema exactly.
// Triage priority is set separately via PATCH /triage/{visit_id}/priority.
export type VitalsCreate = {
  temperature: number | null
  bp_systolic: number | null
  bp_diastolic: number | null
  heart_rate: number | null
  oxygen_saturation: number | null
  respiratory_rate: number | null
  weight_kg: number | null
}

export const DUPLICATE_VITALS_MESSAGE =
  'Vitals have already been recorded for this visit.'
